sql/tests: panic in sqlstats

sql/tests.TestRandomSyntaxFunctions [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3290922&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3290922&tab=artifacts#/) on master @ [e723a85739a2e1b514f5de3221a2177a3a37d0e6](https://github.com/cockroachdb/cockroach/commits/e723a85739a2e1b514f5de3221a2177a3a37d0e6):

Fatal error:

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x273d2b9]

```
Stack: 

```
goroutine 1203574 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc009dca280, 0x6864950, 0xc01abe4ae0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:231 +0x126
panic(0x4a61d60, 0x8bd1680)
	/usr/local/go/src/runtime/panic.go:965 +0x1b9
github.com/cockroachdb/cockroach/pkg/sql/sqlstats/persistedsqlstats.(*PersistedSQLStats).doFlushSingleStmtStats.func1(0x6864950, 0xc01abe4ae0, 0xc00b33c370, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlstats/persistedsqlstats/flush.go:130 +0x59
github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn.func1(0x6864950, 0xc01abe4ae0, 0xc00b33c370, 0xc00c67c240, 0x3ff0000000000000)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:794 +0x43
github.com/cockroachdb/cockroach/pkg/kv.(*Txn).exec(0xc00b33c370, 0x6864950, 0xc01abe4ae0, 0xc002b8db68, 0xc00b33c370, 0x11ae1a6a97f)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/txn.go:878 +0xdc
github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn(0xc009dcaf80, 0x6864950, 0xc01abe4ae0, 0xc002b8dbb8, 0xc002b8dbe8, 0x9674a6)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:793 +0x105
github.com/cockroachdb/cockroach/pkg/sql/sqlstats/persistedsqlstats.(*PersistedSQLStats).doFlushSingleStmtStats(0xc00be7a0f0, 0x6864950, 0xc01abe4ae0, 0x0, 0xed8a6a771, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlstats/persistedsqlstats/flush.go:128 +0x85
github.com/cockroachdb/cockroach/pkg/sql/sqlstats/persistedsqlstats.(*PersistedSQLStats).Flush.func1.1(0xc00be7a0f0, 0x290c9ed9)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlstats/persistedsqlstats/flush.go:39 +0x45
github.com/cockroachdb/cockroach/pkg/sql/sqlstats/persistedsqlstats.(*PersistedSQLStats).doFlush(0xc00be7a0f0, 0x6864950, 0xc01abe4ae0, 0xc002b8dd10, 0x5245bbe, 0x24)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlstats/persistedsqlstats/flush.go:73 +0x10a
github.com/cockroachdb/cockroach/pkg/sql/sqlstats/persistedsqlstats.(*PersistedSQLStats).Flush.func1(0x6864950, 0xc01abe4ae0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlstats/persistedsqlstats/flush.go:38 +0x9c
github.com/cockroachdb/cockroach/pkg/sql/sqlstats/sslocal.(*SQLStats).IterateStatementStats(0xc01196ec40, 0x6864950, 0xc01abe4ae0, 0xc01fe9f9ca, 0xc002b8de10, 0x35, 0xc00c81ee20)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlstats/sslocal/sslocal_provider.go:134 +0x93
github.com/cockroachdb/cockroach/pkg/sql/sqlstats/persistedsqlstats.(*PersistedSQLStats).Flush(0xc00be7a0f0, 0x6864950, 0xc01abe4ae0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlstats/persistedsqlstats/flush.go:37 +0x225
github.com/cockroachdb/cockroach/pkg/sql/sqlstats/persistedsqlstats.(*PersistedSQLStats).startSQLStatsFlushLoop.func1(0x6864950, 0xc01abe4ae0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlstats/persistedsqlstats/provider.go:115 +0x24c
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2(0xc009dca280, 0x6864950, 0xc01abe4ae0, 0x0, 0x0, 0xc00c6032a8)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:442 +0xf3
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:434 +0x22b
```
<details><summary>Log preceding fatal error</summary>
<p>

```
    rsg_test.go:736: 25s of 5m0s: 108046 executions, 68505 successful
    rsg_test.go:736: 30s of 5m0s: 128606 executions, 81523 successful
    rsg_test.go:736: 35s of 5m0s: 150284 executions, 95263 successful
    rsg_test.go:736: 40s of 5m0s: 171647 executions, 108751 successful
    rsg_test.go:736: 45s of 5m0s: 193480 executions, 122622 successful
    rsg_test.go:736: 50s of 5m0s: 214918 executions, 136107 successful
    rsg_test.go:736: 55s of 5m0s: 236312 executions, 149678 successful
    rsg_test.go:736: 1m0s of 5m0s: 257902 executions, 163396 successful
E210812 05:23:58.714504 1847606 sql/distsql/server.go:367  [n1,client=127.0.0.1:33510,hostssl,user=root,intExec=read-roles] 1  error setting up flow: root: memory budget exceeded: 10240 bytes requested, 134215680 currently allocated, 134217728 bytes in budget
    rsg_test.go:736: 1m5s of 5m0s: 279174 executions, 176755 successful
E210812 05:24:01.206197 1872224 sql/distsql/server.go:367  [n1,client=127.0.0.1:35612,hostssl,user=root,intExec=read-roles] 2  error setting up flow: root: memory budget exceeded: 10240 bytes requested, 134215680 currently allocated, 134217728 bytes in budget
E210812 05:24:01.206384 1872188 sql/distsql/server.go:367  [n1,client=127.0.0.1:35586,hostssl,user=root,intExec=read-roles] 3  error setting up flow: root: memory budget exceeded: 10240 bytes requested, 134215680 currently allocated, 134217728 bytes in budget
E210812 05:24:01.688734 1203574 1@util/log/logcrash/crash_reporting.go:140  [n1] 4  a panic has occurred!
E210812 05:24:01.688734 1203574 1@util/log/logcrash/crash_reporting.go:140  [n1] 4 +runtime error: invalid memory address or nil pointer dereference
E210812 05:24:01.688734 1203574 1@util/log/logcrash/crash_reporting.go:140  [n1] 4 +(1) attached stack trace
E210812 05:24:01.688734 1203574 1@util/log/logcrash/crash_reporting.go:140  [n1] 4 +  -- stack trace:
E210812 05:24:01.688734 1203574 1@util/log/logcrash/crash_reporting.go:140  [n1] 4 +  | runtime.gopanic
E210812 05:24:01.688734 1203574 1@util/log/logcrash/crash_reporting.go:140  [n1] 4 +  | 	/usr/local/go/src/runtime/panic.go:965
E210812 05:24:01.688734 1203574 1@util/log/logcrash/crash_reporting.go:140  [n1] 4 +  | runtime.panicmem
E210812 05:24:01.688734 1203574 1@util/log/logcrash/crash_reporting.go:140  [n1] 4 +  | 	/usr/local/go/src/runtime/panic.go:212
E210812 05:24:01.688734 1203574 1@util/log/logcrash/crash_reporting.go:140  [n1] 4 +  | runtime.sigpanic
E210812 05:24:01.688734 1203574 1@util/log/logcrash/crash_reporting.go:140  [n1] 4 +  | 	/usr/local/go/src/runtime/signal_unix.go:734
E210812 05:24:01.688734 1203574 1@util/log/logcrash/crash_reporting.go:140  [n1] 4 +  | github.com/cockroachdb/cockroach/pkg/sql/sqlstats/persistedsqlstats.(*PersistedSQLStats).doFlushSingleStmtStats.func1
E210812 05:24:01.688734 1203574 1@util/log/logcrash/crash_reporting.go:140  [n1] 4 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlstats/persistedsqlstats/flush.go:130
E210812 05:24:01.688734 1203574 1@util/log/logcrash/crash_reporting.go:140  [n1] 4 +  | github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn.func1
E210812 05:24:01.688734 1203574 1@util/log/logcrash/crash_reporting.go:140  [n1] 4 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:794
E210812 05:24:01.688734 1203574 1@util/log/logcrash/crash_reporting.go:140  [n1] 4 +  | github.com/cockroachdb/cockroach/pkg/kv.(*Txn).exec
E210812 05:24:01.688734 1203574 1@util/log/logcrash/crash_reporting.go:140  [n1] 4 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/kv/txn.go:878
E210812 05:24:01.688734 1203574 1@util/log/logcrash/crash_reporting.go:140  [n1] 4 +  | github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn
E210812 05:24:01.688734 1203574 1@util/log/logcrash/crash_reporting.go:140  [n1] 4 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:793
E210812 05:24:01.688734 1203574 1@util/log/logcrash/crash_reporting.go:140  [n1] 4 +  | github.com/cockroachdb/cockroach/pkg/sql/sqlstats/persistedsqlstats.(*PersistedSQLStats).doFlushSingleStmtStats
E210812 05:24:01.688734 1203574 1@util/log/logcrash/crash_reporting.go:140  [n1] 4 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlstats/persistedsqlstats/flush.go:128
E210812 05:24:01.688734 1203574 1@util/log/logcrash/crash_reporting.go:140  [n1] 4 +  | github.com/cockroachdb/cockroach/pkg/sql/sqlstats/persistedsqlstats.(*PersistedSQLStats).Flush.func1.1
E210812 05:24:01.688734 1203574 1@util/log/logcrash/crash_reporting.go:140  [n1] 4 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlstats/persistedsqlstats/flush.go:39
E210812 05:24:01.688734 1203574 1@util/log/logcrash/crash_reporting.go:140  [n1] 4 +  | github.com/cockroachdb/cockroach/pkg/sql/sqlstats/persistedsqlstats.(*PersistedSQLStats).doFlush
E210812 05:24:01.688734 1203574 1@util/log/logcrash/crash_reporting.go:140  [n1] 4 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlstats/persistedsqlstats/flush.go:73
E210812 05:24:01.688734 1203574 1@util/log/logcrash/crash_reporting.go:140  [n1] 4 +  | github.com/cockroachdb/cockroach/pkg/sql/sqlstats/persistedsqlstats.(*PersistedSQLStats).Flush.func1
E210812 05:24:01.688734 1203574 1@util/log/logcrash/crash_reporting.go:140  [n1] 4 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlstats/persistedsqlstats/flush.go:38
E210812 05:24:01.688734 1203574 1@util/log/logcrash/crash_reporting.go:140  [n1] 4 +  | github.com/cockroachdb/cockroach/pkg/sql/sqlstats/sslocal.(*SQLStats).IterateStatementStats
E210812 05:24:01.688734 1203574 1@util/log/logcrash/crash_reporting.go:140  [n1] 4 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlstats/sslocal/sslocal_provider.go:134
E210812 05:24:01.688734 1203574 1@util/log/logcrash/crash_reporting.go:140  [n1] 4 +  | github.com/cockroachdb/cockroach/pkg/sql/sqlstats/persistedsqlstats.(*PersistedSQLStats).Flush
E210812 05:24:01.688734 1203574 1@util/log/logcrash/crash_reporting.go:140  [n1] 4 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlstats/persistedsqlstats/flush.go:37
E210812 05:24:01.688734 1203574 1@util/log/logcrash/crash_reporting.go:140  [n1] 4 +  | github.com/cockroachdb/cockroach/pkg/sql/sqlstats/persistedsqlstats.(*PersistedSQLStats).startSQLStatsFlushLoop.func1
E210812 05:24:01.688734 1203574 1@util/log/logcrash/crash_reporting.go:140  [n1] 4 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlstats/persistedsqlstats/provider.go:115
E210812 05:24:01.688734 1203574 1@util/log/logcrash/crash_reporting.go:140  [n1] 4 +  | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2
E210812 05:24:01.688734 1203574 1@util/log/logcrash/crash_reporting.go:140  [n1] 4 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:442
E210812 05:24:01.688734 1203574 1@util/log/logcrash/crash_reporting.go:140  [n1] 4 +  | runtime.goexit
E210812 05:24:01.688734 1203574 1@util/log/logcrash/crash_reporting.go:140  [n1] 4 +  | 	/usr/local/go/src/runtime/asm_amd64.s:1371
E210812 05:24:01.688734 1203574 1@util/log/logcrash/crash_reporting.go:140  [n1] 4 +Wraps: (2) runtime error: invalid memory address or nil pointer dereference
E210812 05:24:01.688734 1203574 1@util/log/logcrash/crash_reporting.go:140  [n1] 4 +Error types: (1) *withstack.withStack (2) runtime.errorString
```
</p>
</details>
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestRandomSyntaxFunctions PKG=./pkg/sql/tests TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #66514 sql/tests: TestRandomSyntaxFunctions failed [C-test-failure O-robot branch-release-21.1]
- #63943 sql/tests: TestRandomSyntaxFunctions failed [C-test-failure O-robot T-sql-experience branch-release-20.2]
</p>
</details>
<details><summary>Internal log</summary>
<p>

```
mjibson marked as alumn{us/a}; resolving to rafiss instead
```
</p>
</details>

/cc @cockroachdb/sql-experience rafiss
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRandomSyntaxFunctions.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/BUILD.bazel
+++ b/pkg/BUILD.bazel
@@ -298,6 +298,7 @@ ALL_TESTS = [
     "//pkg/sql/sqlliveness/slstorage:slstorage_test",
     "//pkg/sql/sqlstats/persistedsqlstats/sqlstatsutil:sqlstatsutil_test",
     "//pkg/sql/sqlstats/persistedsqlstats:persistedsqlstats_test",
+    "//pkg/sql/sqlstats/sslocal:sslocal_test",
     "//pkg/sql/stats:stats_test",
     "//pkg/sql/stmtdiagnostics:stmtdiagnostics_test",
     "//pkg/sql/tests:tests_test",

--- a/pkg/sql/sqlstats/sslocal/BUILD.bazel
+++ b/pkg/sql/sqlstats/sslocal/BUILD.bazel
@@ -1,4 +1,4 @@
-load("@io_bazel_rules_go//go:def.bzl", "go_library")
+load("@io_bazel_rules_go//go:def.bzl", "go_library", "go_test")
 
 go_library(
     name = "sslocal",
@@ -28,5 +28,28 @@ go_library(
         "//pkg/util/syncutil",
         "//pkg/util/timeutil",
         "@com_github_cockroachdb_errors//:errors",
+    ],
+)
+
+go_test(
+    name = "sslocal_test",
+    srcs = [
+        "iterator_test.go",
+        "main_test.go",
+    ],
+    deps = [
+        "//pkg/base",
+        "//pkg/roachpb:with-mocks",
+        "//pkg/security",
+        "//pkg/security/securitytest",
+        "//pkg/server",
+        "//pkg/sql",
+        "//pkg/sql/sqlstats",
+        "//pkg/testutils/serverutils",
+        "//pkg/testutils/sqlutils",
+        "//pkg/testutils/testcluster",
+        "//pkg/util/leaktest",
+        "//pkg/util/log",
+        "@com_github_stretchr_testify//require",
     ],
 )

--- a/pkg/sql/sqlstats/sslocal/iterator_test.go
+++ b/pkg/sql/sqlstats/sslocal/iterator_test.go
@@ -1,0 +1,94 @@
+// Copyright 2021 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package sslocal_test
+
+import (
+	"context"
+	"testing"
+
+	"github.com/cockroachdb/cockroach/pkg/base"
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/sql"
+	"github.com/cockroachdb/cockroach/pkg/sql/sqlstats"
+	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
+	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
+	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/stretchr/testify/require"
+)
+
+func TestSQLStatsIteratorWithTelemetryFlush(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	ctx := context.Background()
+	s, goDB, _ := serverutils.StartServer(t, base.TestServerArgs{})
+	defer s.Stopper().Stop(ctx)
+
+	testCases := map[string]string{
+		"SELECT _":    "SELECT 1",
+		"SELECT _, _": "SELECT 1, 1",
+	}
+
+	sqlConn := sqlutils.MakeSQLRunner(goDB)
+
+	for _, stmt := range testCases {
+		sqlConn.Exec(t, stmt)
+	}
+
+	sqlStats := s.SQLServer().(*sql.Server).GetSQLStatsProvider()
+
+	// We collect all the statement fingerprint IDs so that we can test the
+	// transaction stats later.
+	fingerprintIDs := make(map[roachpb.StmtFingerprintID]struct{})
+	require.NoError(t,
+		sqlStats.IterateStatementStats(ctx, &sqlstats.IteratorOptions{},
+			func(_ context.Context, statistics *roachpb.CollectedStatementStatistics) error {
+				fingerprintIDs[statistics.ID] = struct{}{}
+				return nil
+			}))
+
+	t.Run("statement_iterator", func(t *testing.T) {
+		require.NoError(t,
+			sqlStats.IterateStatementStats(
+				ctx,
+				&sqlstats.IteratorOptions{},
+				func(_ context.Context, statistics *roachpb.CollectedStatementStatistics) error {
+					require.NotNil(t, statistics)
+					// If we are running our test case, we reset the SQL Stats. The iterator
+					// should gracefully handle that.
+					if _, ok := testCases[statistics.Key.Query]; ok {
+						require.NoError(t, sqlStats.Reset(ctx))
+					}
+					return nil
+				}))
+	})
+
+	t.Run("transaction_iterator", func(t *testing.T) {
+		for _, stmt := range testCases {
+			sqlConn.Exec(t, stmt)
+		}
+		require.NoError(t,
+			sqlStats.IterateTransactionStats(
+				ctx,
+				&sqlstats.IteratorOptions{},
+				func(ctx context.Context, id roachpb.TransactionFingerprintID, statistics *roachpb.CollectedTransactionStatistics) error {
+					require.NotNil(t, statistics)
+
+					for _, stmtFingerprintID := range statistics.StatementFingerprintIDs {
+						if _, ok := fingerprintIDs[stmtFingerprintID]; ok {
+							require.NoError(t, sqlStats.Reset(ctx))
+						}
+					}
+					return nil
+				}))
+	})
+}

--- a/pkg/sql/sqlstats/sslocal/main_test.go
+++ b/pkg/sql/sqlstats/sslocal/main_test.go
@@ -1,0 +1,29 @@
+// Copyright 2021 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package sslocal_test
+
+import (
+	"os"
+	"testing"
+
+	"github.com/cockroachdb/cockroach/pkg/security"
+	"github.com/cockroachdb/cockroach/pkg/security/securitytest"
+	"github.com/cockroachdb/cockroach/pkg/server"
+	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
+	"github.com/cockroachdb/cockroach/pkg/testutils/testcluster"
+)
+
+func TestMain(m *testing.M) {
+	security.SetAssetLoader(securitytest.EmbeddedAssets)
+	serverutils.InitTestServerFactory(server.TestServerFactory)
+	serverutils.InitTestClusterFactory(testcluster.TestClusterFactory)
+	os.Exit(m.Run())
+}

--- a/pkg/sql/sqlstats/ssmemstorage/ss_mem_iterator.go
+++ b/pkg/sql/sqlstats/ssmemstorage/ss_mem_iterator.go
@@ -26,7 +26,8 @@ type baseIterator struct {
 // inside of a ssmemstorage.Container.
 type StmtStatsIterator struct {
 	baseIterator
-	stmtKeys stmtList
+	stmtKeys     stmtList
+	currentValue *roachpb.CollectedStatementStatistics
 }
 
 // NewStmtStatsIterator returns a StmtStatsIterator.
@@ -52,16 +53,14 @@ func NewStmtStatsIterator(
 	}
 }
 
-// Next increments the internal counter of the StmtStatsIterator. It returns
-// true if the following Cur() call is valid, false otherwise.
+// Next updates the current value returned by the subsequent Cur() call. Next()
+// returns true if the following Cur() call is valid, false otherwise.
 func (s *StmtStatsIterator) Next() bool {
 	s.idx++
-	return s.idx < len(s.stmtKeys)
-}
+	if s.idx >= len(s.stmtKeys) {
+		return false
+	}
 
-// Cur returns the roachpb.CollectedStatementStatistics at the current internal
-// counter.
-func (s *StmtStatsIterator) Cur() *roachpb.CollectedStatementStatistics {
 	stmtKey := s.stmtKeys[s.idx]
 
 	stmtFingerprintID := constructStatementFingerprintIDFromStmtKey(stmtKey)
@@ -70,9 +69,9 @@ func (s *StmtStatsIterator) Cur() *roachpb.CollectedStatementStatistics {
 
 	// If the key is not found (and we expected to find it), the table must
 	// have been cleared between now and the time we read all the keys. In
-	// that case we simply skip this key as there are no metrics to report.
+	// that case we simply skip this key and keep advancing the iterator.
 	if statementStats == nil {
-		return nil
+		return s.Next()
 	}
 
 	statementStats.mu.Lock()
@@ -83,7 +82,7 @@ func (s *StmtStatsIterator) Cur() *roachpb.CollectedStatementStatistics {
 	database := statementStats.mu.database
 	statementStats.mu.Unlock()
 
-	collectedStats := roachpb.CollectedStatementStatistics{
+	s.currentValue = &roachpb.CollectedStatementStatistics{
 		Key: roachpb.StatementStatisticsKey{
 			Query:       stmtKey.anonymizedStmt,
 			DistSQL:     distSQLUsed,
@@ -99,14 +98,22 @@ func (s *StmtStatsIterator) Cur() *roachpb.CollectedStatementStatistics {
 		Stats: data,
 	}
 
-	return &collectedStats
+	return true
+}
+
+// Cur returns the roachpb.CollectedStatementStatistics at the current internal
+// counter.
+func (s *StmtStatsIterator) Cur() *roachpb.CollectedStatementStatistics {
+	return s.currentValue
 }
 
 // TxnStatsIterator is an iterator that iterates over the transaction statistics
-// inside of a ssmemstorage.Container.
+// inside a ssmemstorage.Container.
 type TxnStatsIterator struct {
 	baseIterator
-	txnKeys txnList
+	txnKeys  txnList
+	curValue *roachpb.CollectedTransactionStatistics
+	curKey   roachpb.TransactionFingerprintID
 }
 
 // NewTxnStatsIterator returns a new instance of TxnStatsIterator.
@@ -132,11 +139,39 @@ func NewTxnStatsIterator(
 	}
 }
 
-// Next increments the internal counter of the TxnStatsIterator. It returns
-// true if the following Cur() call is valid, false otherwise.
+// Next updates the current value returned by the subsequent Cur() call. Next()
+// returns true if the following Cur() call is valid, false otherwise.
 func (t *TxnStatsIterator) Next() bool {
 	t.idx++
-	return t.idx < len(t.txnKeys)
+	if t.idx >= len(t.txnKeys) {
+		return false
+	}
+
+	txnKey := t.txnKeys[t.idx]
+
+	// We don't want to create the key if it doesn't exist, so it's okay to
+	// pass nil for the statementFingerprintIDs, as they are only set when a key is
+	// constructed.
+	txnStats, _, _ := t.container.getStatsForTxnWithKey(txnKey, nil /* stmtFingerprintIDs */, false /* createIfNonexistent */)
+
+	// If the key is not found (and we expected to find it), the table must
+	// have been cleared between now and the time we read all the keys. In
+	// that case we simply skip this key and advance the iterator.
+	if txnStats == nil {
+		return t.Next()
+	}
+
+	txnStats.mu.Lock()
+	defer txnStats.mu.Unlock()
+
+	t.curKey = txnKey
+	t.curValue = &roachpb.CollectedTransactionStatistics{
+		StatementFingerprintIDs: txnStats.statementFingerprintIDs,
+		App:                     t.container.appName,
+		Stats:                   txnStats.mu.data,
+	}
+
+	return true
 }
 
 // Cur returns the roachpb.CollectedTransactionStatistics at the current internal
@@ -145,25 +180,5 @@ func (t *TxnStatsIterator) Cur() (
 	roachpb.TransactionFingerprintID,
 	*roachpb.CollectedTransactionStatistics,
 ) {
-	txnKey := t.txnKeys[t.idx]
-
-	// We don't want to create the key if it doesn't exist, so it's okay to
-	// pass nil for the statementFingerprintIDs, as they are only set when a key is
-	// constructed.
-	txnStats, _, _ := t.container.getStatsForTxnWithKey(txnKey, nil /* stmtFingerprintIDs */, false /* createIfNonexistent */)
-	// If the key is not found (and we expected to find it), the table must
-	// have been cleared between now and the time we read all the keys. In
-	// that case we simply skip this key as there are no metrics to report.
-	if txnStats == nil {
-		return 0, nil
-	}
-
-	txnStats.mu.Lock()
-	defer txnStats.mu.Unlock()
-	collectedStats := roachpb.CollectedTransactionStatistics{
-		StatementFingerprintIDs: txnStats.statementFingerprintIDs,
-		App:                     t.container.appName,
-		Stats:                   txnStats.mu.data,
-	}
-	return txnKey, &collectedStats
+	return t.curKey, t.curValue
 }

